From d113abcb411e311e3a0b03532930eff4e3273bbd Mon Sep 17 00:00:00 2001
From: Anjaneyulu Reddy Mule <anjaneyulu.reddy.mule@xilinx.com>
Date: Fri, 28 Aug 2020 17:02:43 +0530
Subject: [PATCH] Kernel patch for separated RQ PI, CQ CI DB memory

Signed-off-by: Anjaneyulu Reddy Mule <anjaneyulu.reddy.mule@xilinx.com>
---
 include/uapi/rdma/xib-abi.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/uapi/rdma/xib-abi.h b/include/uapi/rdma/xib-abi.h
index a278334..bdaabaa 100755
--- a/include/uapi/rdma/xib-abi.h
+++ b/include/uapi/rdma/xib-abi.h
@@ -65,6 +65,9 @@ struct xib_ib_create_qp_resp {
 struct xib_ib_alloc_ucontext_resp {
 	__aligned_u64 db_pa;
 	__u32 db_size;
+	__aligned_u64 cq_ci_db_pa;
+	__aligned_u64 rq_pi_db_pa;
+	uint32_t cq_ci_db_size, rq_pi_db_size;
 	__u32 qp_tab_size;
 };
 
-- 
2.7.4

